spatialdata/_core/transform.py

```diff
@@ -1,86 +1,526 @@
-from functools import singledispatch
-from typing import Any, Optional
+from __future__ import annotations
+
+import copy
+import json
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Type, Union
 
-import anndata as ad
 import numpy as np
-import xarray as xr
 
 from spatialdata._types import ArrayLike
 
+__all__ = [
+    "BaseTransformation",
+    "Identity",
+    "MapIndex",
+    "MapAxis",
+    "Translation",
+    "Scale",
+    "Affine",
+    "Rotation",
+    "Sequence",
+    "Displacements",
+    "Coordinates",
+    "VectorField",
+    "InverseOf",
+    "Bijection",
+    "ByDimension",
+    "compose_transformations",
+]
+
+
+class BaseTransformation(ABC):
+    @property
+    @abstractmethod
+    def ndim(self) -> Optional[int]:
+        pass
+
+    @abstractmethod
+    def to_dict(self) -> Dict[str, Any]:
+        pass
+
+    def to_json(self) -> str:
+        return json.dumps(self.to_dict())
+
+    def compose_with(self, transformation: BaseTransformation) -> BaseTransformation:
+        return compose_transformations(self, transformation)
+
+    @abstractmethod
+    def transform_points(self, points: ArrayLike) -> ArrayLike:
+        pass
+
+    @abstractmethod
+    def inverse(self) -> BaseTransformation:
+        pass
+
+
+def compose_transformations(transformation0: BaseTransformation, transformation1: BaseTransformation) -> Sequence:
+    return Sequence([transformation0, transformation1])
+
+
+def get_transformation_from_json(s: str) -> BaseTransformation:
```

Comment:
nitpick, would make these `def _compose...` and would move to end of file.

Reply:
I actually removed since it was redundant: if private, so not exposed to the user, then `Compose` does the same job (and more).